// @dcloudio/uni-app-uts/dist/plugins/android/uvue/compiler/transforms/transformStyle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformStyle = void 0;
const uni_nvue_styler_1 = require("@dcloudio/uni-nvue-styler");
const errors_1 = require("../errors");
// Verify that the template style is in compliance with specifications
const transformStyle = (node, context) => {
    if (node.type === 1 /* NodeTypes.ELEMENT */) {
        node.props.forEach((p, i) => {
            if (p.type === 6 /* NodeTypes.ATTRIBUTE */ && p.name === 'style' && p.value) {
                (0, uni_nvue_styler_1.parse)(p.value.content, {
                    logLevel: 'WARNING',
                    map: true,
                    ts: true,
                    noCode: true,
                    type: 'uvue',
                    platform: process.env.UNI_UTS_PLATFORM,
                }).then(({ messages }) => {
                    messages.forEach((message) => {
                        context.onWarn((0, errors_1.createCompilerError)(100, p.loc, {
                            100: message.text,
                        }));
                    });
                });
            }
        });
    }
};
exports.transformStyle = transformStyle;
