// @dcloudio/uni-app-uts/dist/plugins/android/uvue/compiler/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.addEasyComponentAutoImports = exports.genComponentPublicInstanceImported = exports.genComponentPublicInstanceIdent = exports.genImportComponentPublicInstance = exports.onCompilerError = exports.rewriteObjectExpression = exports.genRenderFunctionDecl = void 0;
const path_1 = __importDefault(require("path"));
const shared_1 = require("@vue/shared");
const compiler_core_1 = require("@vue/compiler-core");
const compiler_sfc_1 = require("@vue/compiler-sfc");
const parser_1 = require("@babel/parser");
const uni_cli_shared_1 = require("@dcloudio/uni-cli-shared");
const transformExpression_1 = require("./transforms/transformExpression");
const utils_1 = require("../../utils");
function genRenderFunctionDecl({ className }) {
    return `function ${className}Render(): VNode | null`;
}
exports.genRenderFunctionDecl = genRenderFunctionDecl;
function rewriteObjectExpression(exp, context) {
    const source = (0, transformExpression_1.stringifyExpression)(exp);
    if (source.includes('{')) {
        const s = new compiler_sfc_1.MagicString(source);
        const ast = (0, parser_1.parseExpression)(source, {
            plugins: context.expressionPlugins,
        });
        (0, compiler_sfc_1.walk)(ast, {
            enter(node) {
                if (node.type === 'ObjectExpression') {
                    s.prependLeft(node.start, node.properties.length > 0
                        ? 'utsMapOf('
                        : 'utsMapOf<string, any | null>(');
                    s.prependRight(node.end, ')');
                }
            },
        });
        return (0, compiler_core_1.createSimpleExpression)(s.toString(), false, exp.loc);
    }
}
exports.rewriteObjectExpression = rewriteObjectExpression;
function onCompilerError(error) { }
exports.onCompilerError = onCompilerError;
function genImportComponentPublicInstance(rootDir, tagName, fileName) {
    if (fileName.includes('@dcloudio')) {
        return '';
    }
    return `, { ${genComponentPublicInstanceImported(rootDir, fileName)} as ${genComponentPublicInstanceIdent(tagName)} }`;
}
exports.genImportComponentPublicInstance = genImportComponentPublicInstance;
function genComponentPublicInstanceIdent(tagName) {
    return (0, shared_1.capitalize)((0, shared_1.camelize)(tagName)) + 'ComponentPublicInstance';
}
exports.genComponentPublicInstanceIdent = genComponentPublicInstanceIdent;
function genComponentPublicInstanceImported(root, fileName) {
    if (path_1.default.isAbsolute(fileName)) {
        fileName = (0, uni_cli_shared_1.normalizePath)(path_1.default.relative(root, fileName));
    }
    if (fileName.startsWith('@/')) {
        return (0, utils_1.genClassName)(fileName.replace('@/', '')) + 'ComponentPublicInstance';
    }
    return (0, utils_1.genClassName)(fileName) + 'ComponentPublicInstance';
}
exports.genComponentPublicInstanceImported = genComponentPublicInstanceImported;
function addEasyComponentAutoImports(easyComponentAutoImports, rootDir, tagName, fileName) {
    // 内置easycom，如 unicloud-db
    if (fileName.includes('@dcloudio')) {
        return;
    }
    if (path_1.default.isAbsolute(fileName)) {
        fileName = '@/' + (0, uni_cli_shared_1.normalizePath)(path_1.default.relative(rootDir, fileName));
    }
    easyComponentAutoImports[fileName] = [
        genComponentPublicInstanceImported(rootDir, fileName),
        genComponentPublicInstanceIdent(tagName),
    ];
}
exports.addEasyComponentAutoImports = addEasyComponentAutoImports;
